File: apps/ios/Tests/CrossyUITests/RoomWeatherTests.swift
import Foundation
import XCTest

import CrossyStore

@testable import CrossyUI

// Honest weather (apps/ios/DESIGN.md §8; the three states are PROTOCOL.md §7's):
// live is a calm dot, resyncing is a breathing dot AND NOTHING ELSE (the board
// keeps its last truth until the snapshot lands), reconnecting dims the room with
// a quiet countdown. Connecting (client-local, pre-first-welcome) dims without a
// countdown. What changes on the board versus the chrome per state is exactly what
// these tests pin.

final class RoomWeatherTests: XCTestCase {
    func test_live_isACalmDotAndNothingChanges_design8() {
        let weather = RoomWeather.from(sync: .live)
        XCTAssertEqual(weather.dot, .calm)
        XCTAssertFalse(weather.boardDimmed)
        XCTAssertFalse(weather.showsCountdown)
        XCTAssertNil(weather.label)
    }

    func test_resyncing_breathesOnChromeOnly_theBoardHoldsItsTruth_protocol7() {
        let weather = RoomWeather.from(sync: .resyncing)
        XCTAssertEqual(weather.dot, .breathing)
        // The board does not dim: sequenced state stays rendered until the sync
        // snapshot replaces it wholesale.
        XCTAssertFalse(weather.boardDimmed)
        XCTAssertFalse(weather.showsCountdown)
        XCTAssertNil(weather.label)
    }

    func test_reconnecting_dimsTheRoomWithAQuietCountdown_design8() {
        let weather = RoomWeather.from(sync: .reconnecting)
        XCTAssertEqual(weather.dot, .dimmed)
        XCTAssertTrue(weather.boardDimmed)
        XCTAssertTrue(weather.showsCountdown)
        XCTAssertEqual(weather.label, "Reconnecting")
    }

    // The FIRST connect is the terse, quiet register (redesign 2026-07-11, DESIGN.md
    // §8: the room's law is a hush, never a spinner). A reconnect names itself and
    // counts down because a person LOST something mid-solve; a first join has lost
    // nothing, so the pill carries no word and no countdown, only the dimmed dot beside
    // the clock. The board still dims (honestly not live yet), but the pill stays quiet,
    // so its width never snaps narrow when the welcome lands.
    func test_connecting_isTheTerseQuietRegister_noWordNoCountdown_design8() {
        let weather = RoomWeather.from(sync: .connecting)
        XCTAssertEqual(weather.dot, .dimmed)
        XCTAssertTrue(weather.boardDimmed)
        XCTAssertFalse(weather.showsCountdown, "a first join has lost nothing to count toward")
        XCTAssertNil(weather.label, "the terse first-connect pill carries no status word")
    }

    // A first connect and a reconnect diverge honestly: only the reconnect names itself
    // and counts down (the pill's width change on reconnect is then a real event, not a
    // first-frame guess). The distinction is pinned so the two registers never re-merge.
    func test_firstConnectAndReconnectDivergeOnTheWordAndTheCountdown_design8() {
        let first = RoomWeather.from(sync: .connecting)
        let again = RoomWeather.from(sync: .reconnecting)
        XCTAssertNil(first.label)
        XCTAssertFalse(first.showsCountdown)
        XCTAssertEqual(again.label, "Reconnecting")
        XCTAssertTrue(again.showsCountdown)
        XCTAssertEqual(first.dot, again.dot, "both are the dimmed dot; the word is the difference")
    }

    func test_countdownSeconds_ceilsAndFloorsAtZero() {
        let now = Date(timeIntervalSinceReferenceDate: 1000)
        XCTAssertNil(RoomWeather.countdownSeconds(retryAt: nil, now: now))
        XCTAssertEqual(
            RoomWeather.countdownSeconds(retryAt: now.addingTimeInterval(2.4), now: now), 3)
        XCTAssertEqual(
            RoomWeather.countdownSeconds(retryAt: now.addingTimeInterval(-5), now: now), 0)
    }

    func test_reconnectLine_isPlainAndWarm_id5() {
        let now = Date(timeIntervalSinceReferenceDate: 1000)
        XCTAssertEqual(
            RoomWeather.reconnectLine(retryAt: now.addingTimeInterval(2.4), now: now),
            "Back in 3s")
        // No deadline, or one already passed, reads as the bare state word: the
        // room never promises a number it does not have.
        XCTAssertEqual(RoomWeather.reconnectLine(retryAt: nil, now: now), "Reconnecting")
        XCTAssertEqual(
            RoomWeather.reconnectLine(retryAt: now.addingTimeInterval(-1), now: now),
            "Reconnecting")
    }

    func test_boardDim_isAWashNotABlackout() {
        // The room never dims dead (DESIGN.md §4): the wash must leave the grid
        // clearly readable behind it.
        XCTAssertLessThan(RoomWeather.boardDimOpacity, 0.6)
        XCTAssertGreaterThan(RoomWeather.boardDimOpacity, 0.2)
    }
}

// The reconnect overlay grace (Track A-ios, PROTOCOL.md §7): the non-live pair
// (resyncing, reconnecting) surfaces its weather only after the connection has
// been continuously non-live for RoomWeather.reconnectOverlayGraceSeconds, and
// hides the instant it recovers. Railway's edge recycles a healthy socket on a
// schedule and reconnect-and-resync heals in ~200 ms, so a bare recycle must
// never flash the overlay. Pure and clock-injected, pinned headlessly here.
final class ReconnectOverlayGateTests: XCTestCase {
    private let epoch = Date(timeIntervalSinceReferenceDate: 1000)
    private var grace: Double { RoomWeather.reconnectOverlayGraceSeconds }

    func test_graceMatchesTheWebConstant_2000ms_protocol7() {
        // The iOS window is the twin of the web's RECONNECT_OVERLAY_GRACE_MS.
        XCTAssertEqual(RoomWeather.reconnectOverlayGraceSeconds, 2)
    }

    func test_nonLivePairIsWithheldUntilTheGraceElapses_protocol7() {
        var gate = ReconnectOverlayGate()
        gate.observe(.reconnecting, now: epoch)
        // Just short of the window: still presented as live (no overlay).
        XCTAssertFalse(gate.overlayPresented(now: epoch.addingTimeInterval(grace - 0.001)))
        XCTAssertEqual(
            gate.presentedSync(.reconnecting, now: epoch.addingTimeInterval(grace - 0.001)), .live)
        // At the window it surfaces its true register.
        XCTAssertTrue(gate.overlayPresented(now: epoch.addingTimeInterval(grace)))
        XCTAssertEqual(
            gate.presentedSync(.reconnecting, now: epoch.addingTimeInterval(grace)), .reconnecting)
    }

    func test_resyncingIsGatedByTheSameWindow_protocol7() {
        var gate = ReconnectOverlayGate()
        gate.observe(.resyncing, now: epoch)
        XCTAssertEqual(
            gate.presentedSync(.resyncing, now: epoch.addingTimeInterval(grace - 0.5)), .live)
        XCTAssertEqual(
            gate.presentedSync(.resyncing, now: epoch.addingTimeInterval(grace)), .resyncing)
    }

    func test_recoveryHidesTheOverlayImmediately_protocol7() {
        var gate = ReconnectOverlayGate()
        gate.observe(.reconnecting, now: epoch)
        // Recover well past the window: live clears the origin, so the overlay is
        // gone at once with no lingering grace.
        gate.observe(.live, now: epoch.addingTimeInterval(grace + 5))
        XCTAssertNil(gate.nonLiveSince)
        XCTAssertFalse(gate.overlayPresented(now: epoch.addingTimeInterval(grace + 5)))
        XCTAssertEqual(gate.presentedSync(.live, now: epoch.addingTimeInterval(grace + 5)), .live)
    }

    func test_bounceBetweenTheNonLivePairSharesOneOrigin_noFlicker_protocol7() {
        var gate = ReconnectOverlayGate()
        gate.observe(.resyncing, now: epoch)
        // A gap that becomes a drop mid-grace: the origin does NOT move, so the
        // window is measured from the first non-live instant, never restarted.
        gate.observe(.reconnecting, now: epoch.addingTimeInterval(1.5))
        XCTAssertEqual(gate.nonLiveSince, epoch)
        // Still withheld a hair before the ORIGINAL deadline.
        XCTAssertEqual(
            gate.presentedSync(.reconnecting, now: epoch.addingTimeInterval(grace - 0.001)), .live)
        // Surfaces at the original deadline, in its current register.
        XCTAssertEqual(
            gate.presentedSync(.reconnecting, now: epoch.addingTimeInterval(grace)), .reconnecting)
    }

    func test_connectingIsNeverGated_theFirstConnectKeepsItsQuietRegister_protocol7() {
        var gate = ReconnectOverlayGate()
        // A first connect passes through untouched (a first join has lost nothing,
        // DESIGN.md §8); it never stamps a non-live origin.
        gate.observe(.connecting, now: epoch)
        XCTAssertNil(gate.nonLiveSince)
        XCTAssertEqual(gate.presentedSync(.connecting, now: epoch.addingTimeInterval(grace)), .connecting)
    }

    func test_secondsUntilPresented_isTheRemainingWindowThenNil() throws {
        var gate = ReconnectOverlayGate()
        XCTAssertNil(gate.secondsUntilPresented(now: epoch), "live has nothing to wake for")
        gate.observe(.reconnecting, now: epoch)
        XCTAssertEqual(
            try XCTUnwrap(gate.secondsUntilPresented(now: epoch)), grace, accuracy: 0.0001)
        XCTAssertEqual(
            try XCTUnwrap(gate.secondsUntilPresented(now: epoch.addingTimeInterval(1.5))),
            grace - 1.5, accuracy: 0.0001)
        // Once the window has elapsed the overlay already shows, so there is
        // nothing left to schedule.
        XCTAssertNil(gate.secondsUntilPresented(now: epoch.addingTimeInterval(grace)))
    }
}

// The chrome model owns the grace-gated presentation the room feeds into
// RoomWeather.from (Track A-ios). Clock-injected so the flip is pinned without a
// real sleep; the one-shot wake it arms merely re-runs this same recompute at the
// deadline. Presentation only: the store's SyncState is never touched.
@MainActor
final class ReconnectGraceModelTests: XCTestCase {
    private let epoch = Date(timeIntervalSinceReferenceDate: 1000)
    private var grace: Double { RoomWeather.reconnectOverlayGraceSeconds }

    func test_presentedSyncWithholdsThePairUntilTheGraceElapses_protocol7() {
        let chrome = RoomChromeModel()
        chrome.observeReconnectGrace(.live, now: epoch)
        XCTAssertEqual(chrome.presentedSync, .live)
        // Drop: within the window the room still reads live (no overlay, input on).
        chrome.observeReconnectGrace(.reconnecting, now: epoch)
        XCTAssertEqual(chrome.presentedSync, .live)
        chrome.observeReconnectGrace(.reconnecting, now: epoch.addingTimeInterval(grace - 0.5))
        XCTAssertEqual(chrome.presentedSync, .live)
        // At the window the true register surfaces.
        chrome.observeReconnectGrace(.reconnecting, now: epoch.addingTimeInterval(grace))
        XCTAssertEqual(chrome.presentedSync, .reconnecting)
    }

    func test_recoveryRevertsPresentedSyncImmediately_protocol7() {
        let chrome = RoomChromeModel()
        chrome.observeReconnectGrace(.reconnecting, now: epoch)
        chrome.observeReconnectGrace(.reconnecting, now: epoch.addingTimeInterval(grace))
        XCTAssertEqual(chrome.presentedSync, .reconnecting)
        // Welcome lands: back to live at once, and the pending wake is cancelled.
        chrome.observeReconnectGrace(.live, now: epoch.addingTimeInterval(grace + 0.2))
        XCTAssertEqual(chrome.presentedSync, .live)
    }

    func test_bounceKeepsOneSharedTimer_noFlicker_protocol7() {
        let chrome = RoomChromeModel()
        chrome.observeReconnectGrace(.resyncing, now: epoch)
        // A resyncing-to-reconnecting bounce mid-grace stays withheld against the
        // ORIGINAL origin, never restarting the window.
        chrome.observeReconnectGrace(.reconnecting, now: epoch.addingTimeInterval(1.5))
        XCTAssertEqual(chrome.presentedSync, .live)
        chrome.observeReconnectGrace(.reconnecting, now: epoch.addingTimeInterval(grace))
        XCTAssertEqual(chrome.presentedSync, .reconnecting)
    }
}

File: apps/android/ui/src/test/kotlin/crossy/ui/RoomWeatherTests.kt
// The honest-weather mapping pinned against apps/ios/Sources/CrossyUI/RoomWeather.swift (DESIGN.md
// §8): which states dim the board, which count down, and the "Back in Ns" line. The board-dim wash
// opacity and the countdown math are the behavior the room bar and the grid scrim read; keeping them
// pinned keeps the Android twin from drifting off the iOS numbers.
package crossy.ui

import crossy.store.SyncState
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class RoomWeatherTests {
    @Test
    fun `DESIGN8 the board dims pre-welcome and while reconnecting, never while live or resyncing`() {
        assertTrue(RoomWeather.boardDimmed(SyncState.CONNECTING))
        assertTrue(RoomWeather.boardDimmed(SyncState.RECONNECTING))
        assertFalse(RoomWeather.boardDimmed(SyncState.LIVE))
        assertFalse(RoomWeather.boardDimmed(SyncState.RESYNCING))
    }

    @Test
    fun `DESIGN8 the board-dim wash matches the iOS 0_45 opacity`() {
        assertEquals(0.45, RoomWeather.boardDimOpacity)
    }

    @Test
    fun `DESIGN8 only reconnecting counts down, a first connect names nothing`() {
        assertTrue(RoomWeather.showsCountdown(SyncState.RECONNECTING))
        assertFalse(RoomWeather.showsCountdown(SyncState.CONNECTING))
        assertFalse(RoomWeather.showsCountdown(SyncState.LIVE))
        assertFalse(RoomWeather.showsCountdown(SyncState.RESYNCING))
    }

    @Test
    fun `DESIGN8 whole seconds to the next dial, ceiled and floored at zero`() {
        assertNull(RoomWeather.countdownSeconds(retryAtMillis = null, nowMillis = 1_000))
        assertEquals(3, RoomWeather.countdownSeconds(retryAtMillis = 4_000, nowMillis = 1_000))
        // Ceils a partial second (iOS rounds up), and never goes negative once the deadline passes.
        assertEquals(3, RoomWeather.countdownSeconds(retryAtMillis = 3_500, nowMillis = 1_000))
        assertEquals(0, RoomWeather.countdownSeconds(retryAtMillis = 900, nowMillis = 1_000))
    }

    @Test
    fun `DESIGN8 the countdown line is warm and plain, the bare word once it elapses`() {
        assertEquals("Back in 3s", RoomWeather.reconnectLine(retryAtMillis = 4_000, nowMillis = 1_000))
        assertEquals("Reconnecting", RoomWeather.reconnectLine(retryAtMillis = null, nowMillis = 1_000))
        assertEquals("Reconnecting", RoomWeather.reconnectLine(retryAtMillis = 1_000, nowMillis = 1_000))
    }

    @Test
    fun `DESIGN8 the dot's three registers, calm live, breathing resync, dimmed while not live`() {
        assertEquals(RoomWeather.Dot.CALM, RoomWeather.dot(SyncState.LIVE))
        assertEquals(RoomWeather.Dot.BREATHING, RoomWeather.dot(SyncState.RESYNCING))
        assertEquals(RoomWeather.Dot.DIMMED, RoomWeather.dot(SyncState.CONNECTING))
        assertEquals(RoomWeather.Dot.DIMMED, RoomWeather.dot(SyncState.RECONNECTING))
    }

    @Test
    fun `DESIGN8 only a reconnect names itself, the first connect and the calm states stay wordless`() {
        assertEquals("Reconnecting", RoomWeather.label(SyncState.RECONNECTING))
        assertNull(RoomWeather.label(SyncState.CONNECTING))
        assertNull(RoomWeather.label(SyncState.LIVE))
        assertNull(RoomWeather.label(SyncState.RESYNCING))
    }

    // The reconnect-overlay grace window (Track A-android; the web and iOS 2000 ms twin): the chrome
    // waits out RECONNECT_OVERLAY_GRACE_MS of continuous non-live before it speaks, and clears the
    // instant the room returns to live. Driven on the test scheduler (the SessionDriverTests idiom).
    @OptIn(ExperimentalCoroutinesApi::class)
    @Test
    fun `A-android a non-live blink under the grace window never shows the overlay`() = runTest {
        val sync = MutableStateFlow(SyncState.LIVE)
        val seen = mutableListOf<Boolean>()
        val job = launch { RoomWeather.overlayGrace(sync).collect { seen.add(it) } }
        runCurrent()

        sync.value = SyncState.RECONNECTING
        advanceTimeBy(RoomWeather.RECONNECT_OVERLAY_GRACE_MS - 1)
        runCurrent()

        // A ~200 ms Railway reconnect recovers before the window elapses: the chrome never spoke.
        assertEquals(listOf(false), seen)
        job.cancel()
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    @Test
    fun `A-android non-live past the grace window shows the overlay`() = runTest {
        val sync = MutableStateFlow(SyncState.LIVE)
        val seen = mutableListOf<Boolean>()
        val job = launch { RoomWeather.overlayGrace(sync).collect { seen.add(it) } }
        runCurrent()

        sync.value = SyncState.RECONNECTING
        advanceTimeBy(RoomWeather.RECONNECT_OVERLAY_GRACE_MS)
        runCurrent()

        assertEquals(listOf(false, true), seen)
        job.cancel()
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    @Test
    fun `A-android recovery to live hides the overlay at once`() = runTest {
        val sync = MutableStateFlow(SyncState.LIVE)
        val seen = mutableListOf<Boolean>()
        val job = launch { RoomWeather.overlayGrace(sync).collect { seen.add(it) } }
        runCurrent()

        sync.value = SyncState.RECONNECTING
        advanceUntilIdle() // past the window: the overlay is up
        assertEquals(listOf(false, true), seen)

        sync.value = SyncState.LIVE
        runCurrent() // no virtual time passes: the overlay clears immediately on recovery
        assertEquals(listOf(false, true, false), seen)
        job.cancel()
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    @Test
    fun `A-android recovery before the window cancels the pending timer`() = runTest {
        val sync = MutableStateFlow(SyncState.LIVE)
        val seen = mutableListOf<Boolean>()
        val job = launch { RoomWeather.overlayGrace(sync).collect { seen.add(it) } }
        runCurrent()

        sync.value = SyncState.RESYNCING
        advanceTimeBy(RoomWeather.RECONNECT_OVERLAY_GRACE_MS - 1)
        runCurrent()
        sync.value = SyncState.LIVE // recovered before the window: the armed timer is cancelled
        advanceUntilIdle() // let any stranded timer fire; it must not

        assertEquals(listOf(false), seen)
        job.cancel()
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    @Test
    fun `A-android a resyncing to reconnecting bounce is one unbroken non-live stretch`() = runTest {
        val sync = MutableStateFlow(SyncState.LIVE)
        val seen = mutableListOf<Boolean>()
        val job = launch { RoomWeather.overlayGrace(sync).collect { seen.add(it) } }
        runCurrent()

        sync.value = SyncState.RESYNCING
        advanceTimeBy(RoomWeather.RECONNECT_OVERLAY_GRACE_MS / 2)
        runCurrent()
        sync.value = SyncState.RECONNECTING // a bounce mid-window must not reset the shared timer
        advanceTimeBy(RoomWeather.RECONNECT_OVERLAY_GRACE_MS / 2)
        runCurrent()

        // The window is measured from the first non-live instant, not the bounce: the overlay shows.
        assertEquals(listOf(false, true), seen)
        job.cancel()
    }
}
